sympy/solvers/solveset.py

```diff
@@ -635,11 +634,12 @@ def _solveset(f, symbol, domain, _check=False):
             result = FiniteSet(*[s for s in result
                       if isinstance(s, RootOf)
                       or domain_check(fx, symbol, s)])
-
+    if _was_domain:
+        return Intersection(result, domain)
     return result
```

Comment:
I think the only change that needs to be made is to put the original domain-handling logic here:

```
if domain is S.Complexes or domain is S.Reals:
    return result
return Intersection(result, domain)
```
I think all other changes you have made may be unnecessary.  Does that work?